### PR TITLE
Add zarr3 streaming v0

### LINKED ISSUE #7933: Add zarr3 streaming
## Detailed Description
Webknossos can already read zarr3 data. It would be great to also stream data out. However, the OME-Zarr metadata for zarr3 is not finalized yet. There is a proposal that we can implement right now: https://github.com/ome/ngff/pull/242 (with the version tag "0.5-dev2")
The first iteration doesn't need to have sharding, so it will basically just need some new routes for the metadata. We should put this under an experimental route, e.g. `/data/zarr_experimental/scalable_minds/l4_sample/color/`

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
### URL of deployed dev instance (used for testing):
- https://addzarrstreamingv.webknossos.xyz

### Steps to test:
- Check the zarr.jsons:
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/color/1/zarr.json
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/color/zarr.json
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/segmentation/1/zarr.json
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/segmentation/zarr.json
- Check the listing routes:
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/ 
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/datasource-properties.json
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/color/
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/segmentation/
  - https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/color/2-2-1/
- Check whether an annotation is listed properly:
  - https://addzarrstreamingv.webknossos.xyz/data/annotations/zarr3_experimental/S53FgwKKnbWczqUq/
  - https://addzarrstreamingv.webknossos.xyz/data/annotations/zarr3_experimental/S53FgwKKnbWczqUq/color
  - https://addzarrstreamingv.webknossos.xyz/data/annotations/zarr3_experimental/S53FgwKKnbWczqUq/segmentation
  - https://addzarrstreamingv.webknossos.xyz/data/annotations/zarr3_experimental/S53FgwKKnbWczqUq/datasource-properties.json
  - https://addzarrstreamingv.webknossos.xyz/data/annotations/zarr3_experimental/S53FgwKKnbWczqUq/segmentation/4-4-1/
  - https://addzarrstreamingv.webknossos.xyz/data/annotations/zarr3_experimental/S53FgwKKnbWczqUq/color/4-4-1/
 - The following "remote-local" dataset should render for mags 1&2&4 https://addzarrstreamingv.webknossos.xyz/datasets/sample_organization/test-zarr3-viewing/view#3573,3568,1024,0,2.358
 - Same for the segmentation layer: https://addzarrstreamingv.webknossos.xyz/datasets/sample_organization/Zarr3-Segmentation-Streaming-Test/view#3584,3584,1024,0,1 (only mag 1 is imported)
 - And here I streamed an annotation (all mags): https://addzarrstreamingv.webknossos.xyz/datasets/sample_organization/zarr3-annotation-streaming#3584,3584,1024,0,9.85
 - All of them should work :)

### Issues:
- fixes #7933

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [documentation](../blob/master/docs) if applicable -> maybe?
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change -> maybe add new issue there?
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md) -> @fm3 do you see any cases I did not include in the tests?
- [x] Needs datastore update after deployment

### CODE REVIEW

**Comment by joshmoore** [comment]
This is great! 🙌🏽 When there's a URL that can be tested, please let me know.

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/dataformats/zarr/ZarrCoordinatesParser.scala [hunk 1]
```diff
@@ -4,12 +4,20 @@ object ZarrCoordinatesParser {
...
         Some(Integer.parseInt(c), Integer.parseInt(x), Integer.parseInt(y), Integer.parseInt(z))
       case _ => None
     }
   }
+
+  def parseNDimensionalDotCoordinates(
+      coordinates: String,
+  ): Option[Array[Int]] = {
+    val ndCoordinatesRx = "^\\s*c\\.([0-9]+)\\.([0-9]+)\\.([0-9]+)(\\.([0-9]+))+\\s*$".r
+    // The tail cuts of the leading "c" form the "c." at the beginning of coordinates.
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/dataformats/zarr/ZarrCoordinatesParser.scala:20]:
```suggestion
    // The tail cuts off the leading "c" form the "c." at the beginning of coordinates.
```

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala [hunk 2]
```diff
@@ -174,53 +230,62 @@ class ZarrStreamingController @Inject()(
...
       mag: String,
-      cxyz: String,
+      coordinates: String,
   )(implicit m: MessagesProvider): Fox[Result] =
     for {
       (dataSource, dataLayer) <- dataSourceRepository.getDataSourceAndDataLayer(organizationName,
                                                                                 datasetName,
                                                                                 dataLayerName) ~> NOT_FOUND
-      (c, x, y, z) <- ZarrCoordinatesParser.parseDotCoordinates(cxyz) ?~> "zarr.invalidChunkCoordinates" ~> NOT_FOUND
+      //  (c, x, y, z)
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala:275]: remove?

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala [hunk 3]
```diff
@@ -174,53 +230,62 @@ class ZarrStreamingController @Inject()(
...
+      parsedCoordinates <- ZarrCoordinatesParser.parseNDimensionalDotCoordinates(coordinates) ?~> "zarr.invalidChunkCoordinates" ~> NOT_FOUND // TODO: change error message
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala:276]: todo

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala [hunk 4]
```diff
@@ -174,53 +230,62 @@ class ZarrStreamingController @Inject()(
...
       magParsed <- Vec3Int.fromMagLiteral(mag, allowScalar = true) ?~> Messages("dataLayer.invalidMag", mag) ~> NOT_FOUND
       _ <- bool2Fox(dataLayer.containsResolution(magParsed)) ?~> Messages("dataLayer.wrongMag", dataLayerName, mag) ~> NOT_FOUND
-      _ <- bool2Fox(c == 0) ~> "zarr.invalidFirstChunkCoord" ~> NOT_FOUND
+      _ <- bool2Fox(parsedCoordinates.head == 0) ~> "zarr.invalidFirstChunkCoord" ~> NOT_FOUND
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala:279]: head can throw an exception. access with headOption instead
```suggestion
      channelCoordinate <- parsedcoordinates.headOption ?~> NOT_FOUND
      _ <- bool2Fox(channelCoordinate == 0) ~> "zarr.invalidFirstChunkCoord" ~> NOT_FOUND
```

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala [hunk 5]
```diff
@@ -174,53 +230,62 @@ class ZarrStreamingController @Inject()(
...
+      (x, y, z) = (parsedCoordinates(parsedCoordinates.length - 3),
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala:280]: this index access can also throw exceptions. please first assert with bool2fox that parsedCoordinates.length is large enough

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala [hunk 6]
```diff
@@ -174,53 +230,62 @@ class ZarrStreamingController @Inject()(
...
+                   parsedCoordinates(parsedCoordinates.length - 2),
+                   parsedCoordinates(parsedCoordinates.length - 1))
+      additionalCoordinates = Some(
+        parsedCoordinates
+          .slice(1, parsedCoordinates.length - 3)
+          .zipWithIndex
+          .map(coordWithIndex => new AdditionalCoordinate(name = s"t${coordWithIndex._2}", value = coordWithIndex._1))
+          .toList)
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/ZarrStreamingController.scala:288]: could be extracted to a function? Or maybe even be part of the coordinate parsing?

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr/NgffMetadata.scala [hunk 7]
```diff
@@ -57,6 +58,22 @@ object NgffMultiscalesItem {
   implicit val jsonFormat: OFormat[NgffMultiscalesItem] = Json.format[NgffMultiscalesItem]
 }
 
+case class NgffMultiscalesItemV2(
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr/NgffMetadata.scala:61]: I’d say the v2 stuff should be in separate file. If some code is shared between v1 and v2, maybe add a third file for the shared stuff.

Also, is v2 the right version number? I thought they were like at 0.5?

Also, please add some links here to the spec documents.

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr3/Zarr3ArrayHeader.scala [hunk 8]
```diff
@@ -248,5 +249,41 @@ object Zarr3ArrayHeader extends JsonImplicits {
         "storage_transformers" -> zarrArrayHeader.storage_transformers,
         "dimension_names" -> zarrArrayHeader.dimension_names
       )
+
+  }
+  def fromDataLayerToVersion3(dataLayer: DataLayer): Zarr3ArrayHeader = {
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr3/Zarr3ArrayHeader.scala:254]: since this is a member of object Zarr3ArrayHeader, I’d say `fromDataLayer` is sufficient as a name, what do you think?

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr3/Zarr3GroupHeader.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      groupHeaderBuilder ++= Seq(
+        "zarr_format" -> zarr3GroupHeader.zarr_format,
+        "node_type" -> zarr3GroupHeader.node_type,
+      )
+      if (zarr3GroupHeader.ngffMetadata.isDefined) {
+        groupHeaderBuilder += ("attributes" -> Json.obj("ome" -> zarr3GroupHeader.ngffMetadata.get))
+      }
+      groupHeaderBuilder.result()
+    }
+  }
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr3/Zarr3GroupHeader.scala:41]: is there a reason for the custom reads and writes? can the generated ones from Json.format not handle some of the data in here? If so, could you add a brief comment explaining it?

**Review by fm3 — COMMENTED** [review]
Looking pretty good already :) I added a couple comments. Will do some testing next

File: webknossos-datastore/conf/com.scalableminds.webknossos.datastore.routes [hunk 10]
```diff
@@ -42,6 +42,17 @@ GET           /annotations/zarr/:accessTokenOrId/:dataLayerName/:mag/
...
 
+# Zarr3 compatible routes
+GET           /zarr3_experimental/:organizationName/:datasetName/:dataLayerName/zarr.json                                                     @com.scalableminds.webknossos.datastore.controllers.ZarrStreamingController.requestZarrJson(token: Option[String], organizationName: String, datasetName: String, dataLayerName: String)
+GET           /zarr3_experimental/:organizationName/:datasetName/:dataLayerName/:mag/zarr.json                                                @com.scalableminds.webknossos.datastore.controllers.ZarrStreamingController.requestZarrJsonForMag(token: Option[String], organizationName: String, datasetName: String, dataLayerName: String, mag: String)
+GET           /zarr3_experimental/:organizationName/:datasetName/:dataLayerName/:mag/:coordinates                                             @com.scalableminds.webknossos.datastore.controllers.ZarrStreamingController.requestRawZarrCube(token: Option[String], organizationName: String, datasetName: String, dataLayerName: String, mag: String, coordinates: String)
+
+
```
Comment (fm3) [inline webknossos-datastore/conf/com.scalableminds.webknossos.datastore.routes:50]:
```suggestion
```

File: webknossos-datastore/conf/com.scalableminds.webknossos.datastore.routes [hunk 11]
```diff
@@ -42,6 +42,17 @@ GET           /annotations/zarr/:accessTokenOrId/:dataLayerName/:mag/
...
+GET           /annotations/zarr3_experimental/:accessTokenOrId/:dataLayerName/zarr.json                                                       @com.scalableminds.webknossos.datastore.controllers.ZarrStreamingController.zarrJsonWithAnnotationPrivateLink(token: Option[String], accessTokenOrId: String, dataLayerName: String)
+GET           /annotations/zarr3_experimental/:accessTokenOrId/:dataLayerName/:mag/zarr.json                                                  @com.scalableminds.webknossos.datastore.controllers.ZarrStreamingController.zarrJsonPrivateLink(token: Option[String], accessTokenOrId: String, dataLayerName: String, mag: String)
+GET           /annotations/zarr3_experimental/:accessTokenOrId/:dataLayerName/:mag/:coordinates                                               @com.scalableminds.webknossos.datastore.controllers.ZarrStreamingController.rawZarrCubePrivateLink(token: Option[String], accessTokenOrId: String, dataLayerName: String, mag: String, coordinates: String)
+
+
```
Comment (fm3) [inline webknossos-datastore/conf/com.scalableminds.webknossos.datastore.routes:55]:
```suggestion
```

**Review by fm3 — COMMENTED** [review]
@normanrz do you think we should also already add a datasource-properties.json route and directory listing routes? Will some clients need that?

**Comment by normanrz** [comment]
> @normanrz do you think we should also already add a datasource-properties.json route and directory listing routes? Will some clients need that?

Yes, I think that would be useful.

**Comment by MichaelBuessemeyer** [comment]
This is the how the view of a existing color layer looks like when listing an annotation. It says that the layer belongs to a organization and certain dataset. It this alright or should the title be changed? (Bc it kinda leeks the organization & dataset name (although this is already likely leaked via some properties sent to the client)

image

**Comment by MichaelBuessemeyer** [comment]
> Yes, I think that would be useful.

It should be done. See the added routes in the testing guide.

**Comment by normanrz** [comment]
I think the metadata is not correct yet. For example, https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/color/zarr.json should be similar to this:
```json
{
  "zarr_format": 3,
  "node_type": "group",
  "attributes": {
    "ome": {
      "version": "0.5-dev2",
      "multiscales": [{ ... }]
     }
   }
}
```

You can validate with https://deploy-preview-36--ome-ngff-validator.netlify.app/?source=https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/color

**Comment by MichaelBuessemeyer** [comment]
Thanks for poiting this out @normanrz 
the newest version (not yet deployed) should incorporate the schema fix. 

But the validator seems broken to me. It shows some warnings about not being able to find schema definitions.
image

A auick look showed that the referenced schema json for the version is no longer available under the link linked in the schema definition of https://raw.githubusercontent.com/normanrz/ngff/spec-rfc2/latest/schemas/image.schema 

Thus, I just used a different json validator and with the assumption that `"0.5-dev2"` is the correct version string, the schema of the https://addzarrstreamingv.webknossos.xyz/data/zarr3_experimental/sample_organization/l4_sample/color/zarr.json  should now be valid :)

**Comment by normanrz** [comment]
Now it is green!
Screenshot 2024-08-13 at 16 33 53

**Comment by MichaelBuessemeyer** [comment]
> Now it is green!

Awesome :tada: :rocket:

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr3/Zarr3GroupHeader.scala (hunk 9 above)
Comment (normanrz) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr3/Zarr3GroupHeader.scala:32]: ping

**Review by normanrz — APPROVED** [review]
This looks very nice! Thanks.

I just pinged you again on this one comment by @fm3.

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr3/Zarr3GroupHeader.scala (hunk 9 above)
Comment (MichaelBuessemeyer) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/datareaders/zarr3/Zarr3GroupHeader.scala:32]: The manual json serialization and deserialization is needed to have the correct json format. The default formatter would not put the metadata under "attribbutes.ome" but under "ngffMetadata" as thats how the member of `Zarr3GroupHeader` is called. Therefore the manual serialization is necessary here :) I added comments about this

**Comment by normanrz** [comment]
:shipit:
